### PR TITLE
Add readOnlyInputs option for CharliecloudBuilder

### PR DESCRIPTION
**Pull request by l-modolo**
This merge request is to add the `readOnlyInputs` option as discussed in the  **Charliecloud and shared cacheDir** #3367 issue  and following the  **Add readOnlyInputs option for CharliecloudBuilder** #3477 merge request

With this modification:

- the default `readOnlyInputs` value is `false`
- if `readOnlyInputs` is `true`;
- the `-w` option is removed from the command
- the bound volumes are transformed to only mount the first folder of the path

For example, if I try to bind `/home/user/workdir/` to `/home/` the bound path will be `/home` to `/home`. By doing that, charliecloud doesn't try to `mkdir -p /home/user/workdir/` within the read-only container, the external path is bound at the `/home` directory level so all the subfolders are accessible with the user rights. If one tries to bind `/` I added an expection.

Signed-off-by: Laurent Modolo [laurent.modolo@ens-lyon.fr](mailto:laurent.modolo@ens-lyon.fr)

### CODE REVIEW

File: modules/nextflow/src/main/groovy/nextflow/container/CharliecloudBuilder.groovy [hunk 1]
```diff
@@ -78,9 +84,23 @@ class CharliecloudBuilder extends ContainerBuilder<CharliecloudBuilder> {
         return this
     }
 
+    protected String getRoot(String path) {
+        def rootPath = path.split("/")
+
+        if (rootPath.size() >= 1)
+            rootPath = "/${rootPath[1]}"
```
Comment (pditommaso) [inline modules/nextflow/src/main/groovy/nextflow/container/CharliecloudBuilder.groovy:91]: Not sure about this. What is expected to take? if you want the first path component it should be `rootPath[0]`

Comment (l-modolo) [reply modules/nextflow/src/main/groovy/nextflow/container/CharliecloudBuilder.groovy:91]: Yes, I just want the first path component of the path. However, those are absolute path so `rootPath[0]` is empty

```
println "/test/path".split("/")
[, test, path]
```

Comment (pditommaso) [reply modules/nextflow/src/main/groovy/nextflow/container/CharliecloudBuilder.groovy:91]: use `.tokenize('/')` instead, should avoid that problem

Comment (l-modolo) [reply modules/nextflow/src/main/groovy/nextflow/container/CharliecloudBuilder.groovy]: I made the modification in the commit  a5df098ff

Comment (pditommaso) [reply modules/nextflow/src/main/groovy/nextflow/container/CharliecloudBuilder.groovy]: ok, wonder tho if there's a unit test for it ?

Comment (l-modolo) [reply modules/nextflow/src/main/groovy/nextflow/container/CharliecloudBuilder.groovy]: You are right, I added two test for that in the commit 45de7b8c1

Comment (pditommaso) [reply modules/nextflow/src/main/groovy/nextflow/container/CharliecloudBuilder.groovy]: thanks a lot

**Comment by pditommaso** [comment]
Included in version 23.02.0-edge

**Pull request merged**